projects/observability/src/pages/explorer/explorer-service.test.ts

```diff
@@ -0,0 +1,74 @@
+import { fakeAsync, tick } from '@angular/core/testing';
+import { NavigationParamsType } from '@hypertrace/common';
+import {
+  FilterAttributeType,
+  FilterBuilderLookupService,
+  FilterOperator,
+  NumberFilterBuilder,
+  StringFilterBuilder
+} from '@hypertrace/components';
+import { AttributeMetadataType, MetadataService } from '@hypertrace/distributed-tracing';
+import { ExplorerService, ScopeQueryParam } from '@hypertrace/observability';
+import { createServiceFactory, mockProvider } from '@ngneat/spectator/jest';
+import { of } from 'rxjs';
+
+describe('Explorer service', () => {
+  const createService = createServiceFactory({
+    service: ExplorerService
+  });
+
+  const getAttributeMock = jest.fn((scope, attributeKey) =>
+    of(
+      attributeKey === 'duration'
+        ? {
+            name: 'duration',
+            displayName: 'Duration',
+            units: 'ms',
+            type: AttributeMetadataType.Number,
+            scope: scope,
+            requiresAggregation: false,
+            allowedAggregations: [],
+            groupable: false
+          }
+        : {
+            name: 'userIdentifier',
+            displayName: 'User ID',
+            units: '',
+            type: AttributeMetadataType.String,
+            scope: scope,
+            requiresAggregation: false,
+            allowedAggregations: [],
+            groupable: false
+          }
+    )
+  );
+
+  const lookupMock = jest.fn(filterAttributeType =>
+    filterAttributeType === FilterAttributeType.Number ? new NumberFilterBuilder() : new StringFilterBuilder()
```

Comment:
Follow on from my previous comment - we should either use the real filter builder service if we want to use the real builders and let it handle the details, or mock out the filter builders returned too.

Reply:
The real filter builder service gets injected objects for filter builders and registers them in the service. Mocking the numbers/string filter builders here would make the test irrelevant since we'd be mocking the filter construction logic itself. And the real filter builder service cannot be mocked without injecting all the necessary dependencies. What are the alternatives?

Reply:
> Mocking the numbers/string filter builders here would make the test irrelevant since we'd be mocking the filter construction logic itself.

The idea of abstracting the construction of the filter in this service means we don't care what the construction logic is - just that it gets there, is invoked with the right args, and comes back with the corresponding result. So wouldn't mock a specific filter builder type and its logic, but instead just have lookup return a filter builder mock that has no specific implementation - such as returning "RESULT", which you can validate, as well as validating it was called with the appropriate arguments.

> And the real filter builder service cannot be mocked without injecting all the necessary dependencies. What are the alternatives?

Importing `FilterModule` would handle registering any builders for you. It would be tying the test to other implementations though - not explicitly like using string/number directly, but implicitly such that if they were to change the format of the filter, the test would break.

Reply:
@aaron-steinfeld Done